src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,260 @@
+package ui.components.pickers;
+
+import backend.resource.TurboLabel;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to represent the state of the label picker. In addition to representing a state,
+ * it contains the logic that handles every state transition
+ */
+public class LabelPickerState {
+
+    private Set<String> initialLabels;
+    private List<String> addedLabels;
+    private List<String> removedLabels;
+    private List<String> matchedLabels;
+    private List<TurboLabel> allLabels;
+    private OptionalInt currentSuggestionIndex;
+
+    public LabelPickerState(Set<String> initialLabels, List<TurboLabel> allLabels, String userInput) {
+        this(initialLabels, new ArrayList<>(), new ArrayList<>(), new ArrayList<>(), allLabels,
+                OptionalInt.empty());
+        update(userInput);
+    }
+
+    private LabelPickerState(Set<String> initialLabels, List<String> addedLabels, List<String> removedLabels,
+                             List<String> matchedLabels, List<TurboLabel> allLabels, 
+                             OptionalInt currentSuggestionIndex) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.allLabels = allLabels;
+        this.currentSuggestionIndex = currentSuggestionIndex;
+    }
+
+    /**
+     * Updates current state based on given user input
+     * @param userInput 
+     */
+    private final void update(String userInput) {
+        List<String> confirmedKeywords = getConfirmedKeywords(userInput);
+        for (String confirmedKeyword : confirmedKeywords) {
+            updateIfMatchesLabel(confirmedKeyword);
+        }
+
+        Optional<String> keywordInProgess = getKeywordInProgress(userInput);
+        if (keywordInProgess.isPresent()) {
+            updateMatchedLabels(keywordInProgess.get());
+        }
+    }
+
+    /**
+     * Updates current state if there is at least one matching label based on the 
+     * given keyword
+     * @param keyword
+     */
+    private final void updateIfMatchesLabel(String keyword) {
+        if (TurboLabel.hasMatchedLabel(allLabels, keyword)) {
+            updateAssignedLabels(TurboLabel.getFirstMatchingTurboLabel(allLabels, keyword));
+        }
```

Comment:
Consider the alternative below. Is it more intuitive?:
```java
Optional<TurboLabel> firstMatchingLabel = TurboLabel.getFirstMatchingTurboLabel(allLabels, keyword);
if(firstMatchingLabel.isPresent()) updateAssignedLabels(firstMatchingLabel);
```

In fact, it can be further simplified as follows:
`updateAssignedLabels(TurboLabel.getFirstMatchingTurboLabel(allLabels, keyword));`
In this case, `updateAssignedLabels` should do nothing if the parameter is not present.
If we follow that approach, this method is not even needed.

More generally, if we push the null checking to the lower level methods, those methods become more complete (i.e. they make no assumptions about the parameters) and higher level methods can happily omit null checking. What do you guys think? `Optional` is a new thing, so I myself is not sure of the best practices.